File: 2_semester/Laba_4.2/8v.cpp
#include <iostream>
#include <vector>
using namespace std;

void outdomino(vector<pair<int,int>> &spisok, vector<int> &keys, int n){
	for (auto item : keys){//вывод последовательности
		if (item == -1)
			break;
		cout << spisok[item].first << '/' << spisok[item].second << ' ';
	}
	cout << endl;
}

void revers(vector<pair<int,int>> &a, int n){//переворачивает кость
	int temp = a[n].first;
	a[n].first = a[n].second;
	a[n].second = temp;
}

void domino(vector<pair<int,int>> &spisok, vector<int> &keys, int n){
	bool is_find = false;
	if (n == keys.size()){ //если все кости раставлены, выводим и выходим
		outdomino(spisok, keys, n);
		return;
	}
	for (int i = 0; i < spisok.size(); ++i){
		bool is_repit = false;
		for (auto &item : keys)
			if (item == i){
				is_repit = true;
				break;
			}
		if (is_repit)//если такая уже использовалась, пропускаем шаг
			continue;
		if (n == 0){
			//если мы в первые зашли в рекурсию
			is_find = true;
			keys[0] = i;
			domino(spisok, keys, 1);
			if (spisok[i].first != spisok[i].second){
				revers(spisok, i);
				domino(spisok, keys, 1);
				revers(spisok, i);
			}
			continue;
		}
		if (spisok[i].first == spisok[keys[n-1]].second){ 
			//если 2 номер на прошлой кости совпадает с 1 номером i-ой кости
			keys[n] = i;
			domino(spisok, keys, n+1);
			keys[n] = -1;
			is_find = true;
		}
		if (spisok[i].second == spisok[keys[n-1]].second && spisok[i].second != spisok[i].first){
			//если 2 номер на прошлой кости совпадает с 2 номером i-ой кости и кость не семетрична
			keys[n] = i;
			revers(spisok, i);
			domino(spisok, keys, n+1);
			revers(spisok, i);
			keys[n] = -1;
			is_find = true;
		}
	}
	if (!is_find)
		outdomino(spisok, keys, n);
	return;
}
	
int main(){
	vector<pair<int,int>> spisok = {{1,2}, {2,3}, {5,6}, {2,2}, {5,3}};
	vector<int> keys;
	for(int i = 0; i < spisok.size(); ++i)
		keys.push_back(-1);

	domino(spisok, keys, 0);
}

File: 2_semester/Laba_4.2/8v.py
def outdomino(spisok, keys):
	for i in keys:
		if i == -1:
			break
		print(spisok[i], end = ' ')
	print()

def revers(a, n):
	a[n][0], a[n][1] = a[n][1], a[n][0]

def domino(spisok, keys, n = 0):
	is_find = False
	if n == len(keys):
		outdomino(spisok, keys)
		return
	for i in range(len(spisok)):
		if i in keys:
			continue
		if n == 0:
			is_find = True
			keys[0] = i
			domino(spisok, keys, 1)
			if spisok[i][1] != spisok[i][0]:
				revers(spisok, i)
				domino(spisok, keys, 1)
				revers(spisok, i)
		elif spisok[i][0] == spisok[keys[n-1]][1]: 
			keys[n] = i
			domino(spisok, keys, n+1)
			keys[n] = -1
			is_find = True
		elif spisok[i][1] == spisok[keys[n-1]][1] and spisok[i][1] != spisok[i][0]: 
			keys[n] = i
			revers(spisok, i)
			domino(spisok, keys, n+1)
			revers(spisok, i)
			keys[n] = -1
			is_find = True
	if not is_find:
		outdomino(spisok, keys)
	return

def main():
	spisok = [[1,2], [2,3], [5,6], [2,2], [5,3]]
	keys = [-1 for i in range(len(spisok))]
	domino(spisok, keys)

main()